internal/pkg/cli/file/watch.go

```
@@ -0,0 +1,130 @@
+}
+
+// NewRecursiveWatcher returns a RecursiveWatcher which notifies when changes are made to files inside a recursive directory tree.
+func NewRecursiveWatcher(dir string) (*RecursiveWatcher, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	rw := &RecursiveWatcher{
+		events:          make(chan fsnotify.Event),
+		errors:          make(chan error),
+		fsnotifyWatcher: watcher,
+		done:            make(chan struct{}),
+		closed:          false,
+	}
+
+	go rw.start()
+
+	return rw, nil
+}
+
+// Add recursively adds a directory tree to the list of watched files.
+func (rw *RecursiveWatcher) Add(path string) error {
+	if rw.closed {
+		return fsnotify.ErrClosed
+	}
+	if err := filepath.WalkDir(path, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return rw.fsnotifyWatcher.Add(path)
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Remove recursively removes a directory tree from the list of watched files.
+func (rw *RecursiveWatcher) Remove(path string) error {
+	if rw.closed {
+		return nil
+	}
+	if err := filepath.WalkDir(path, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return rw.fsnotifyWatcher.Remove(path)
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Events returns the events channel.
+func (rw *RecursiveWatcher) Events() <-chan fsnotify.Event {
+	return rw.events
+}
+
+// Errors returns the errors channel.
+func (rw *RecursiveWatcher) Errors() <-chan error {
+	return rw.errors
+}
+
+// Close closes the RecursiveWatcher.
+func (rw *RecursiveWatcher) Close() error {
+	rw.closed = true
+	close(rw.done)
+	return rw.fsnotifyWatcher.Close()
+}
+
+func (rw *RecursiveWatcher) start() {
+	for {
+		select {
+		case event := <-rw.fsnotifyWatcher.Events:
+			info, err := os.Stat(event.Name)
+			if err != nil {
+				break
```

in what scenarios would we get this error? Should we send `err` to `rw.errors`?

---

I think we would only get this error with race conditions caused by editing the file (chmod or rm maybe) after the event is received, which is certainly a rare occurrence. I assumed just swallowing the event is fine because it doesn't mean the watcher itself has failed. Maybe it should just log the error?

---

yeah i like the logging idea. just in case we swallow any error that's useful for debugging